Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -0,0 +1,386 @@
+    std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+    GrpcStreamObserver* observer,
+    GrpcCompletionQueue* grpc_queue) {
+  // `make_shared` requires a public constructor. There are workarounds, but
+  // efficiency is not a big concern here.
+  return std::shared_ptr<GrpcStream>(new GrpcStream{
+      std::move(context), std::move(call), observer, grpc_queue});
+}
+
+void GrpcStream::Start() {
+  HARD_ASSERT(state_ == State::NotStarted, "Stream is already started");
+  state_ = State::Started;
+  Execute<StreamStart>();
+}
+
+void GrpcStream::Read() {
+  HARD_ASSERT(!has_pending_read_,
+              "Cannot schedule another read operation before the previous read "
+              "finishes");
+  HARD_ASSERT(state_ == State::Open, "Read called when the stream is not open");
+
+  has_pending_read_ = true;
+  Execute<StreamRead>();
+}
+
+void GrpcStream::Write(grpc::ByteBuffer&& message) {
+  HARD_ASSERT(state_ == State::Open,
+              "Write called when the stream is not open");
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::Finish() {
+  if (state_ == State::NotStarted) {
+    return;
+  }
+
+  HARD_ASSERT(state_ < State::Finishing, "Finish called twice");
+  state_ = State::Finishing;
+
+  buffered_writer_.reset();
+  // Important: since the stream always has a pending read operation,
+  // cancellation has to be called, or else the read would hang forever, and
+  // finish operation will never get completed (an operation cannot be completed
+  // before all previously-enqueued operations complete).
+  //
+  // On the other hand, when an operation fails, cancellation should not be
+  // called, otherwise the real failure cause will be overwritten by status
+  // "canceled".
+  Execute<ClientInitiatedFinish>();
+}
+
+void GrpcStream::WriteAndFinish(grpc::ByteBuffer&& message) {
+  if (state_ < State::Open) {
+    // Ignore the write part if the call didn't have a chance to open yet.
+    Finish();
+    return;
+  }
+
+  HARD_ASSERT(buffered_writer_,
+              "Write requested when there is no valid buffered_writer_");
+  state_ = State::LastWrite;
+  // Write the last message as soon as possible by discarding anything else that
+  // might be buffered.
+  buffered_writer_->DiscardUnstartedWrites();
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::BufferedWrite(grpc::ByteBuffer&& message) {
+  HARD_ASSERT(buffered_writer_,
+              "Write requested when there is no valid buffered_writer_");
+  StreamWrite* write_operation = MakeOperation<StreamWrite>(std::move(message));
+  buffered_writer_->EnqueueWrite(write_operation);
+}
+
+bool GrpcStream::SameGeneration() const {
+  return generation_ == observer_->generation();
+}
+
+// Callbacks
+
+void GrpcStream::OnStart() {
+  state_ = State::Open;
+  buffered_writer_.emplace(BufferedWriter{});
+
+  if (SameGeneration()) {
+    observer_->OnStreamStart();
+    // Start listening for new messages.
+    Read();
```

Comment:
I find this simple conceptually -- just have a "read" at all times, which completely absolves the caller from all responsibility for calling `Read`. On the other hand, I don't think we expect messages until a listen is called on `WatchStream` or a write method is called on a `WriteStream`:
* do we expect out-of-band messages from the server?
* do you know if having a pending read is somehow wasteful?